File: backend/Ronboard.Api/Services/ClaudeProcessService.cs
namespace Ronboard.Api.Services;

using System.Diagnostics;
using System.Text.Json;
using System.Threading.Channels;
using Ronboard.Api.Models;

public class ClaudeProcessService(ILogger<ClaudeProcessService> logger)
{
    // ── Terminal mode: PTY via `script`, raw byte relay ──

    public (Process Process, ChannelReader<string> Output) StartTerminalProcess(
        string workingDirectory, string? model = null)
    {
        var resolvedDir = ExpandPath(workingDirectory);
        if (!Directory.Exists(resolvedDir))
            throw new DirectoryNotFoundException($"Working directory not found: {resolvedDir}");

        var shell = Environment.GetEnvironmentVariable("SHELL") ?? "/bin/zsh";
        var claudeCmd = "claude";
        if (!string.IsNullOrEmpty(model))
            claudeCmd += $" --model {model}";

        var psi = new ProcessStartInfo
        {
            FileName = "/usr/bin/script",
            Arguments = $"-q /dev/null {shell} -l -c \"{claudeCmd}\"",
            WorkingDirectory = resolvedDir,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };

        psi.Environment["TERM"] = "xterm-256color";
        psi.Environment["COLUMNS"] = "120";
        psi.Environment["LINES"] = "40";

        var process = new Process { StartInfo = psi };
        var channel = Channel.CreateUnbounded<string>(
            new UnboundedChannelOptions { SingleReader = false, SingleWriter = true });

        process.Start();
        logger.LogInformation("Started terminal Claude process (PID {Pid}) in {Dir}", process.Id, resolvedDir);

        _ = Task.Run(async () =>
        {
            try
            {
                var buffer = new char[4096];
                while (true)
                {
                    var count = await process.StandardOutput.ReadAsync(buffer, 0, buffer.Length);
                    if (count == 0) break;
                    await channel.Writer.WriteAsync(new string(buffer, 0, count));
                }
            }
            catch (Exception ex) { logger.LogError(ex, "Error reading terminal stdout"); }
            finally { channel.Writer.Complete(); }
        });

        StartStderrReader(process);
        return (process, channel.Reader);
    }

    public async Task SendInputAsync(Process process, string data)
    {
        await process.StandardInput.WriteAsync(data);
        await process.StandardInput.FlushAsync();
    }

    // ── Stream mode: structured JSON via --print --stream-json ──

    public (Process Process, ChannelReader<ClaudeMessage> Output) StartStreamProcess(
        string workingDirectory, string? model = null)
    {
        var resolvedDir = ExpandPath(workingDirectory);
        if (!Directory.Exists(resolvedDir))
            throw new DirectoryNotFoundException($"Working directory not found: {resolvedDir}");

        var args = "--print --input-format stream-json --output-format stream-json --include-partial-messages --verbose";
        if (!string.IsNullOrEmpty(model))
            args += $" --model {model}";

        var shell = Environment.GetEnvironmentVariable("SHELL") ?? "/bin/zsh";

        var psi = new ProcessStartInfo
        {
            FileName = shell,
            Arguments = $"-l -c \"claude {args}\"",
            WorkingDirectory = resolvedDir,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };

        var process = new Process { StartInfo = psi };
        var channel = Channel.CreateUnbounded<ClaudeMessage>(
            new UnboundedChannelOptions { SingleReader = false, SingleWriter = true });

        process.Start();
        logger.LogInformation("Started stream Claude process (PID {Pid}) in {Dir}", process.Id, resolvedDir);

        _ = Task.Run(async () =>
        {
            var index = 0;
            try
            {
                while (await process.StandardOutput.ReadLineAsync() is { } line)
                {
                    if (string.IsNullOrWhiteSpace(line)) continue;
                    try
                    {
                        var json = JsonDocument.Parse(line);
                        var type = json.RootElement.TryGetProperty("type", out var t)
                            ? t.GetString() ?? "unknown"
                            : "unknown";

                        await channel.Writer.WriteAsync(new ClaudeMessage
                        {
                            Index = index++,
                            Type = type,
                            RawJson = json.RootElement.Clone(),
                        });
                    }
                    catch (JsonException)
                    {
                        logger.LogDebug("Non-JSON output: {Line}", line);
                    }
                }
            }
            catch (Exception ex) { logger.LogError(ex, "Error reading stream stdout"); }
            finally { channel.Writer.Complete(); }
        });

        StartStderrReader(process);
        return (process, channel.Reader);
    }

    public async Task SendStreamMessageAsync(Process process, string userMessage)
    {
        var input = new
        {
            type = "user",
            message = new { role = "user", content = userMessage }
        };
        var json = JsonSerializer.Serialize(input);
        await process.StandardInput.WriteLineAsync(json);
        await process.StandardInput.FlushAsync();
    }

    // ── Shared ──

    private void StartStderrReader(Process process)
    {
        _ = Task.Run(async () =>
        {
            try
            {
                while (await process.StandardError.ReadLineAsync() is { } line)
                {
                    if (!string.IsNullOrWhiteSpace(line))
                        logger.LogWarning("Claude stderr: {Line}", line);
                }
            }
            catch { }
        });
    }

    public static string ExpandPath(string path)
    {
        if (path.StartsWith('~'))
            path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), path[1..].TrimStart('/'));
        return Path.GetFullPath(path);
    }
}

File: backend/Ronboard.Api/ClaudeProcessService.fs
namespace Ronboard.Api.Services

open System
open System.Diagnostics
open System.IO
open System.Text.Json
open System.Threading.Channels
open Microsoft.Extensions.Logging
open Ronboard.Api.Models

type ClaudeProcessService(logger: ILogger<ClaudeProcessService>) =

    let startStderrReader (proc: Process) =
        task {
            try
                let mutable keepReading = true

                while keepReading do
                    let! line = proc.StandardError.ReadLineAsync()

                    if isNull line then
                        keepReading <- false
                    elif not (String.IsNullOrWhiteSpace line) then
                        logger.LogWarning("Claude stderr: {Line}", line)
            with
            | _ -> ()
        }
        |> ignore

    static member ExpandPath(path: string) =
        let p =
            if path.StartsWith('~') then
                Path.Combine(
                    Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                    path.[1..].TrimStart('/')
                )
            else
                path

        Path.GetFullPath(p)

    member _.StartTerminalProcess(workingDirectory: string, ?model: string) =
        let resolvedDir = ClaudeProcessService.ExpandPath(workingDirectory)

        if not (Directory.Exists resolvedDir) then
            raise (DirectoryNotFoundException $"Working directory not found: {resolvedDir}")

        let shell =
            Environment.GetEnvironmentVariable("SHELL")
            |> Option.ofObj
            |> Option.defaultValue "/bin/zsh"

        let claudeCmd =
            match model with
            | Some m -> $"claude --model {m}"
            | None -> "claude"

        let psi =
            ProcessStartInfo(
                FileName = "/usr/bin/script",
                Arguments = $"""-q /dev/null {shell} -l -c "{claudeCmd}" """,
                WorkingDirectory = resolvedDir,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            )

        psi.Environment.["TERM"] <- "xterm-256color"
        psi.Environment.["COLUMNS"] <- "120"
        psi.Environment.["LINES"] <- "40"

        let proc = new Process(StartInfo = psi)

        let channel =
            Channel.CreateUnbounded<string>(
                UnboundedChannelOptions(SingleReader = false, SingleWriter = true)
            )

        proc.Start() |> ignore
        logger.LogInformation("Started terminal Claude process (PID {Pid}) in {Dir}", proc.Id, resolvedDir)

        task {
            try
                let buffer = Array.zeroCreate<char> 4096
                let mutable keepReading = true

                while keepReading do
                    let! count = proc.StandardOutput.ReadAsync(buffer, 0, buffer.Length)

                    if count = 0 then
                        keepReading <- false
                    else
                        do! channel.Writer.WriteAsync(new string (buffer, 0, count))
            with
            | ex -> logger.LogError(ex, "Error reading terminal stdout")

            channel.Writer.Complete()
        }
        |> ignore

        startStderrReader proc
        struct (proc, channel.Reader)

    member _.StartStreamProcess
        (workingDirectory: string, sessionId: Guid, isResume: bool, ?model: string)
        =
        let resolvedDir = ClaudeProcessService.ExpandPath(workingDirectory)

        if not (Directory.Exists resolvedDir) then
            raise (DirectoryNotFoundException $"Working directory not found: {resolvedDir}")

        let mutable args =
            "--print --input-format stream-json --output-format stream-json --include-partial-messages --verbose"

        args <-
            if isResume then
                args + $" --resume {sessionId}"
            else
                args + $" --session-id {sessionId}"

        match model with
        | Some m -> args <- args + $" --model {m}"
        | None -> ()

        let shell =
            Environment.GetEnvironmentVariable("SHELL")
            |> Option.ofObj
            |> Option.defaultValue "/bin/zsh"

        let psi =
            ProcessStartInfo(
                FileName = shell,
                Arguments = $"""-l -c "claude {args}" """,
                WorkingDirectory = resolvedDir,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            )

        let proc = new Process(StartInfo = psi)

        let channel =
            Channel.CreateUnbounded<ClaudeMessage>(
                UnboundedChannelOptions(SingleReader = false, SingleWriter = true)
            )

        proc.Start() |> ignore
        logger.LogInformation("Started stream Claude process (PID {Pid}) in {Dir}", proc.Id, resolvedDir)

        task {
            let mutable index = 0

            try
                let mutable keepReading = true

                while keepReading do
                    let! line = proc.StandardOutput.ReadLineAsync()

                    if isNull line then
                        keepReading <- false
                    elif not (String.IsNullOrWhiteSpace line) then
                        try
                            let json = JsonDocument.Parse(line)

                            let typ =
                                let mutable t = Unchecked.defaultof<JsonElement>

                                if json.RootElement.TryGetProperty("type", &t) then
                                    t.GetString() |> Option.ofObj |> Option.defaultValue "unknown"
                                else
                                    "unknown"

                            do!
                                channel.Writer.WriteAsync(
                                    { Index = index
                                      Timestamp = DateTime.UtcNow
                                      Type = typ
                                      RawJson = json.RootElement.Clone() }
                                )

                            index <- index + 1
                        with
                        | :? JsonException -> logger.LogDebug("Non-JSON output: {Line}", line)
            with
            | ex -> logger.LogError(ex, "Error reading stream stdout")

            channel.Writer.Complete()
        }
        |> ignore

        startStderrReader proc
        struct (proc, channel.Reader)

    member _.SendInputAsync(proc: Process, data: string) =
        task {
            do! proc.StandardInput.WriteAsync(data)
            do! proc.StandardInput.FlushAsync()
        }

    member _.SendStreamMessageAsync(proc: Process, userMessage: string) =
        task {
            let input =
                {| ``type`` = "user"
                   message = {| role = "user"; content = userMessage |} |}

            let json = JsonSerializer.Serialize(input)
            logger.LogInformation("Sending to stream stdin (PID {Pid}): {Json}", proc.Id, json)
            do! proc.StandardInput.WriteLineAsync(json)
            do! proc.StandardInput.FlushAsync()
        }
